### PR TITLE
Add im2text

### PR DESCRIPTION
**Pull request by da03**
Im2Text added. See docs/im2text.md.

### CODE REVIEW

**Comment by da03** [comment]
working to solve travis-ci test failures, do not merge.

**Comment by da03** [comment]
Okay ready to be reviewed @srush

**Comment by srush** [comment]
Looking now. Please don't check in all the files though. Have a binary either on google drive or aws.

File: onmt/IO.py [hunk 1]
```diff
@@ -275,9 +304,14 @@ class ONMTDataset(torchtext.data.Dataset):
     @staticmethod
     def sort_key(ex):
         "Sort in reverse size order"
-        return -len(ex.src)
+        if hasattr(ex, 'src'):
```
Comment (srush) [inline onmt/IO.py]: I don't like the use of `hasattr`, that's a magic method. Please add an actual member to check.

File: onmt/IO.py [hunk 2]
```diff
@@ -249,6 +266,18 @@ def peek(seq):
     return first, chain([first], seq)
 
 
+def _construct_example_fromlist(data, fields):
+    ex = torchtext.data.Example()
+    for (name, field), val in zip(fields, data):
+        if field is not None:
+            if isinstance(val, six.string_types):
```
Comment (srush) [inline onmt/IO.py]: I don't think we need this rstrip thing. Was that from their library?

File: onmt/IO.py [hunk 3]
```diff
@@ -339,28 +384,78 @@ def __init__(self, src_path, tgt_path, fields,
...
+                  for k in keys]
         example_values = ([ex[k] for k in keys] for ex in examples)
-        out_examples = (torchtext.data.Example.fromlist(ex_values, fields)
+        out_examples = (_construct_example_fromlist(ex_values, fields)
                         for ex_values in example_values)
 
         def filter_pred(example):
-            return 0 < len(example.src) <= src_seq_length \
-                and 0 < len(example.tgt) <= tgt_seq_length
+            return (not hasattr(example, 'src')
```
Comment (srush) [inline onmt/IO.py]: Remove `hasattr`.

File: onmt/IO.py [hunk 4]
```diff
@@ -339,28 +384,78 @@ def __init__(self, src_path, tgt_path, fields,
...
+                    or 0 < len(example.tgt) <= tgt_seq_length)
 
         super(ONMTDataset, self).__init__(
             out_examples,
             fields,
-            filter_pred if use_filter_pred else None
+            filter_pred if use_filter_pred else lambda x: True
         )
 
+    def _read_corpus_file(self, path, truncate):
```
Comment (srush) [inline onmt/IO.py]: Where is this used?

File: onmt/ModelConstructor.py [hunk 5]
```diff
@@ -172,10 +172,12 @@ def make_base_model(model_opt, fields, gpu, checkpoint=None):
                 p.data.uniform_(-model_opt.param_init, model_opt.param_init)
             for p in generator.parameters():
                 p.data.uniform_(-model_opt.param_init, model_opt.param_init)
-        model.encoder.embeddings.load_pretrained_vectors(
-                model_opt.pre_word_vecs_enc, model_opt.fix_word_vecs_enc)
-        model.decoder.embeddings.load_pretrained_vectors(
-                model_opt.pre_word_vecs_dec, model_opt.fix_word_vecs_dec)
+        if hasattr(model.encoder, 'embeddings'):
```
Comment (srush) [inline onmt/ModelConstructor.py]: I'll accept the hasattr here.

File: onmt/Translator.py [hunk 6]
```diff
@@ -92,8 +93,12 @@ def translateBatch(self, batch, dataset):
         batch_size = batch.batch_size
 
         # (1) Run the encoder on the src.
-        _, src_lengths = batch.src
-        src = onmt.IO.make_features(batch, 'src')
+        if hasattr(batch, 'src'):
+            _, src_lengths = batch.src
+            src = onmt.IO.make_features(batch, 'src')
+        elif hasattr(batch, 'src_img'):
+            src = onmt.IO.make_features(batch, 'src_img')
```
Comment (srush) [inline onmt/Translator.py]: Why do we need source features here?

File: onmt/modules/ImageEncoder.py [hunk 7]
```diff
@@ -95,8 +94,12 @@ def forward(self, input, lengths=None):
         for row in range(input.size(2)):
             inp = input[:, :, row, :].transpose(0, 2)\
                                      .transpose(1, 2)
-            pos_emb = self.pos_lut(
-                Variable(torch.cuda.LongTensor(batchSize).fill_(row)))
+            if input.is_cuda:
```
Comment (srush) [inline onmt/modules/ImageEncoder.py]: oh, let's just use `type_as(input)` now, instead of an if statement.

File: opts.py [hunk 8]
```diff
@@ -227,6 +227,8 @@ def train_opts(parser):
 def translate_opts(parser):
     parser.add_argument('-model', required=True,
                         help='Path to model .pt file')
+    parser.add_argument('-model_type', default='text',
+                        help="Type of encoder to use. Options are [text|img].")
```
Comment (srush) [inline opts.py]: We should save this with the file. If you need an argument here it should be `data_type`

File: test/test_preprocess.py [hunk 9]
```diff
@@ -17,7 +17,6 @@
 opt.train_tgt = 'data/tgt-train.txt'
 opt.valid_src = 'data/src-val.txt'
 opt.valid_tgt = 'data/tgt-val.txt'
-print(opt)
 
 
 class TestData(unittest.TestCase):
```
Comment (srush) [inline test/test_preprocess.py]: Can you add tests for images too?

File: translate.py [hunk 10]
```diff
@@ -59,9 +59,11 @@ def main():
     if opt.dump_beam != "":
         import json
         translator.initBeamAccum()
-    data = onmt.IO.ONMTDataset(
-        opt.src, opt.tgt, translator.fields,
-        use_filter_pred=False)
+
+    data = onmt.IO.ONMTDataset(opt.model_type,
```
Comment (srush) [inline translate.py]: data_type

**Comment by da03** [comment]
okay, I'll let you know when I resolve those issues.

**Comment by srush** [comment]
@da03 ping.

**Comment by da03** [comment]
okay ready to be reviewed @srush

File: test/test_preprocess.py (hunk 9 above)
Comment (da03) [reply test/test_preprocess.py]: added

**Comment by srush** [comment]
Fantastic! 

I think we will eventually have to clean up IO.py, but for now let's go forward.

**Pull request merged**